Sources/Tools/SoundDecompress/main.cpp

```diff
@@ -0,0 +1,258 @@
+#include "plResMgr/plRegistryHelpers.h"
+#include "plResMgr/plRegistryNode.h"
+
+#include "plAgeDescription/plAgeManifest.h"
+
+#include "plAudioCore/plSoundBuffer.h"
+
+#if OUTPUT_WAV_FILES
+#include "plOldAudioFileReader.h"
+#endif
+
+typedef std::set<std::pair<ST::string, uint16_t>> SoundSet;
+
+enum OutputStyle
+{
+    kSilent,
+    kProgress,
+    kVerbose
+};
+
+//// plSoundBufferCollector //////////////////////////////////////////////////
+//  Page iterator that collects all the plSoundBuffers in all of our pages
+
+class plSoundBufferCollector : public plRegistryPageIterator, public plKeyCollector
+{
+public:
+    plSoundBufferCollector(std::set<plKey>& keyArray)
+                : plKeyCollector(keyArray) {}
+
+    bool EatPage(plRegistryPageNode* page)
+    {
+        if (page->IsValid())
+        {
+            page->LoadKeys();
+            return page->IterateKeys(this, plSoundBuffer::Index());
+        }
+        else
+        {
+            fprintf(stderr, "INVALID PAGE: %s\n", page->GetPagePath().AsString().c_str());
+            return true;
+        }
+    }
+};
+
+
+void PrintHelp()
+{
+    printf("Plasma Sound Decompressor\n");
+    printf("-------------------------\n");
+    printf("-s, --silent        Run silently, no output\n");
+    printf("-v, --verbose       Print each filename when decompressing\n");
+    printf("-f, --force         Force decompressing existing files\n");
+}
+
+bool CollectSounds(plResManager* rm, SoundSet& sfxArray)
+{
+    std::set<plKey> soundKeys;
+
+    plSoundBufferCollector soundCollector(soundKeys);
+    rm->IterateAllPages(&soundCollector);
+
+    for (auto it = soundKeys.begin(); it != soundKeys.end(); ++it)
+    {
+        plSoundBuffer* buffer = plSoundBuffer::ConvertNoRef((*it)->VerifyLoaded());
+        if (buffer)
+        {
+            // Ref it...
+            buffer->GetKey()->RefObject();
+
+            // Get the filename from it and add that file if necessary
+            plFileName filename = buffer->GetFileName();
+            if (filename.IsValid())
+            {
+                uint16_t flags = 0;
+
+                if (filename.GetFileExt().compare_i("wav") != 0)
+                {
+                    if (buffer->HasFlag(plSoundBuffer::kOnlyLeftChannel) ||
+                        buffer->HasFlag(plSoundBuffer::kOnlyRightChannel))
+                        hsSetBits(flags, plManifestFile::kSndFlagCacheSplit);
+                    else if (buffer->HasFlag(plSoundBuffer::kStreamCompressed))
+                        hsSetBits(flags, plManifestFile::kSndFlagStreamCompressed);
+                    else
+                        hsSetBits(flags, plManifestFile::kSndFlagCacheStereo);
+                }
+
+                std::pair<ST::string, uint16_t> pair = std::make_pair(filename.AsString(), flags);
+                sfxArray.insert(pair);
+            }
+
+            // Unref the object so it goes away
+            buffer->GetKey()->UnRefObject();
```

Comment:
Removing these results in the claim that there are a bunch of leaks via plResManager Object Leak Report, although I agree that there should be any leaking because we never Ref the objects.

However, given that there's no downside, I'm going to keep the Ref/UnRef calls since apparently that does a better job of cleaning stuff up

Reply:
That sounds fair for now. Perhaps an issue should be filed about this weird paradigm being needed to prevent leaks?